File: WdtcCore/src/main/java/org/wdt/wdtc/download/fabric/FabricDonwloadInfo.java
package org.wdt.wdtc.download.fabric;


import org.apache.log4j.Logger;
import org.wdt.utils.gson.JSONObject;
import org.wdt.utils.gson.JSONUtils;
import org.wdt.wdtc.download.DownloadTask;
import org.wdt.wdtc.download.infterface.DownloadInfo;
import org.wdt.wdtc.download.infterface.InstallTask;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.manger.FileManger;
import org.wdt.wdtc.utils.WdtcLogger;

import java.io.File;
import java.io.IOException;

public class FabricDonwloadInfo implements DownloadInfo {
    private static final Logger logmaker = WdtcLogger.getLogger(FabricDonwloadInfo.class);
    protected final String FabricVersionNumber;
    protected final Launcher launcher;
    private FabricAPIDownloadTask APIDownloadTask;


    public FabricDonwloadInfo(Launcher launcher, String FabricVersionNumber) {
        this.FabricVersionNumber = FabricVersionNumber;
        this.launcher = launcher;
    }

    public String getFabricVersionNumber() {
        return FabricVersionNumber;
    }


    public String getFabricVersionFileUrl() {
        return Launcher.getOfficialDownloadSource().getFabricMetaUrl() + "v2/versions/loader/%s/%s/profile/json";
    }


    public void DownloadProfileZip() {
        DownloadTask.StartDownloadTask(getProfileZipUrl(), getProfileZipFile());
    }

    public String getProfileZipFile() {
        return String.format(FileManger.getWdtcCache() + "/%s-%s-frofile-zip.zip", launcher.getVersionNumber(), getFabricVersionNumber());
    }

    public String getProfileZipUrl() {
        return String.format(Launcher.getOfficialDownloadSource().getFabricMetaUrl() + "v2/versions/loader/%s/%s/profile/zip", launcher.getVersionNumber(), getFabricVersionNumber());
    }

    public String FromFabricLoaderFolder() {
        return String.format("fabric-loader-%s-%s", getFabricVersionNumber(), launcher.getVersionNumber());
    }

    public File getFabricJar() {
        return new File(FileManger.getWdtcCache() + "/" + FromFabricLoaderFolder() + "/" + FromFabricLoaderFolder() + ".jar");
    }

    public File getFabricVersionJson() {
        return new File(String.format(FileManger.getWdtcCache() + "/%s-fabric-%s.json", launcher.getVersionNumber(), FabricVersionNumber));
    }

    public JSONObject getFabricVersionJsonObject() throws IOException {
        return JSONUtils.getJSONObject(getFabricVersionJson());
    }



    public FabricAPIDownloadTask getAPIDownloadTask() {
        return APIDownloadTask;
    }

    public void setAPIDownloadTask(FabricAPIDownloadTask APIDownloadTask) {
        this.APIDownloadTask = APIDownloadTask;
    }

    public boolean getAPIDownloadTaskNoNull() {
        return APIDownloadTask != null;
    }

    @Override
    public String getModVersion() {
        return FabricVersionNumber;
    }

    @Override
    public InstallTask getModInstallTask() {
        return new FabricInstallTask(launcher, FabricVersionNumber);
    }
}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/download/fabric/FabricDonwloadInfo.kt
package org.wdt.wdtc.core.impl.download.fabric

import com.google.gson.annotations.JsonAdapter
import org.wdt.utils.gson.readFileToClass
import org.wdt.wdtc.core.openapi.download.interfaces.ModDownloadInfoInterface
import org.wdt.wdtc.core.openapi.download.interfaces.ModInstallTaskInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsJsonObjectInterface
import org.wdt.wdtc.core.openapi.game.Version
import org.wdt.wdtc.core.openapi.game.serializeGameRuntimeDependencyGsonBuilder
import org.wdt.wdtc.core.openapi.manger.KindOfMod
import org.wdt.wdtc.core.openapi.manger.officialDownloadSource
import org.wdt.wdtc.core.openapi.manger.wdtcCache
import org.wdt.wdtc.core.openapi.utils.gson.DownloadInfoTypeAdapter
import org.wdt.wdtc.core.openapi.utils.noNull
import org.wdt.wdtc.core.openapi.utils.startDownloadTask
import org.wdt.wdtc.core.openapi.utils.toURL
import java.io.File
import java.net.URL

@JsonAdapter(DownloadInfoTypeAdapter::class)
open class FabricDonwloadInfo(
	protected val version: Version,
	override val modVersion: String
) : ModDownloadInfoInterface {
	var apiDownloadTask: FabricAPIDownloadTask? = null
	
	
	constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) : this(
		version,
		versionsJsonObjectInterface.versionNumber.noNull()
	)
	
	val fabricVersionFileUrl: String
		get() = "${officialDownloadSource.fabricMetaUrl}v2/versions/loader/%s/%s/profile/json"
	
	suspend fun startDownloadProfileZip() {
		startDownloadTask(profileZipUrl, profileZipFile)
	}
	
	
	private val profileZipFile: File
		get() = File(wdtcCache, "${version.versionNumber}-$modVersion-frofile-zip.zip")
	
	private val profileZipUrl: URL
		get() = "${officialDownloadSource.fabricMetaUrl}v2/versions/loader/%s/%s/profile/zip".format(
			version.versionNumber,
			modVersion
		).toURL()
	
	
	private val fromFabricLoaderFolder: String
		get() = "fabric-loader-%s-%s".format(modVersion, version.versionNumber)
	
	
	val fabricJar: File
		get() = File(wdtcCache, "$fromFabricLoaderFolder/$fromFabricLoaderFolder.jar")
	val fabricVersionJson: File
		get() = File(wdtcCache, "${version.versionNumber}-fabric-$modVersion.json")
	
	val fabricProfileJsonObject: FabricProfileJsonObject
		get() = fabricVersionJson.readFileToClass(serializeGameRuntimeDependencyGsonBuilder)
	val isHasApiDownloadTask: Boolean
		get() = apiDownloadTask != null
	override val modInstallTask: ModInstallTaskInterface
		get() = FabricInstallTask(version, modVersion)
	override val modKind: KindOfMod
		get() = KindOfMod.FABRIC
	
	override fun toString(): String {
		return "FabricDonwloadInfo(modVersion='$modVersion', apiDownloadTask=$apiDownloadTask)"
	}
	
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is FabricDonwloadInfo) return false
		
		if (modVersion != other.modVersion) return false
		if (apiDownloadTask != other.apiDownloadTask) return false
		
		return true
	}
	
	override fun hashCode(): Int {
		var result = modVersion.hashCode()
		result = 31 * result + (apiDownloadTask?.hashCode() ?: 0)
		return result
	}
	
	
}
